# Introduction

Data wrangling is the process of cleaning and unifying messy and complex data sets for easy access and analysis and to be familiar with. 
In this project I will gather data from a variety of sources and in a variety of formats, assess its quality and tidiness, then clean it.The dataset that will be wrangling,analyzing and visualizing is the tweet archive of Twitter user @dog_rates, also known as WeRateDogs.

In [47]:
import pandas as pd
import datetime as dt
import numpy as np
import requests
import tweepy
import json
import re
import time
from nltk import pos_tag

# Gather

Gathering Data process for this Project consists of three pieces of data which are:

1- The WeRateDogs Twitter archive file. which was downloaded  manually by clicking the given twitter_archive_enhanced.csv link.

2- The tweet image predictions, This file (image_predictions.tsv) hosted on Udacity's servers and should be downloaded programmatically using the Requests library and the following URL: https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv

3- Each tweet's retweet count and favorite ("like") count at minimum. Using the tweet IDs in the WeRateDogs Twitter archive, I will query the Twitter API for each tweet's JSON data using Python's Tweepy library and store each tweet's entire set of JSON data in a file called tweet_json.txt file. Each tweet's JSON data should be written to its own line. Then I will read this .txt file line by line into a pandas DataFrame with tweet ID, retweet count, and favorite count.


In [93]:
# Read the twitter-archive-enhanced.csv file and store it as dataframe in archive
archive = pd.read_csv('twitter-archive-enhanced.csv')
# Quick check to the file content and structure
archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

The file twitter-archive-enhanced.csv stored in archive data frame with 17 columns and 2356 values.

In [49]:

# Using Requests library to download a file then store it in a tsv file
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)

with open(url.split('/')[-1], mode = 'wb') as file:
    file.write(response.content)

# Read the downloaded file into a dataframe 'images'
images = pd.read_csv('image-predictions.tsv', sep = '\t', encoding = 'utf-8')
# Quick check to the file content and structure
images

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

The file image-predictions.tsv downloaded and stored in images data frame, it has 12 columns and 2075 values.

In [50]:
# Autontification to twetter API

# Generate your own at https://apps.twitter.com/app
# CONSUMER_KEY = 'Consumer Key (API key)'
# CONSUMER_SECRET = 'Consumer Secret (API Secret)'
# OAUTH_TOKEN = 'Access Token'
# OAUTH_TOKEN_SECRET = 'Access Token Secret'

consumer_key = 'qhSwqxpKslaBZofulQi598DrI'
consumer_secret = 'UTdmFpZb24IHcGc8CAaxArwq6LfyVMgUZz1gSuJuHyu7lj7hD0'
access_token = '793618519-K9kFIsgzD9vhp14kD5GxOLnGFEGl3jnjVZnwcMh5'
access_token_secret = 'ySEfCVU4qhRHZ8dTC7r1geQjFABInO0r79MWlK6TZxzLB'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_token_secret)

# Construct the API instance
api = tweepy.API(auth, 
                 parser = tweepy.parsers.JSONParser(), # Parse the result to Json Object
                 wait_on_rate_limit = True, # Automatically wait for rate limits to replenish
                 wait_on_rate_limit_notify = True) # Print a notification when Tweepy is waiting for rate limits to replenish

In [51]:
# Liste where  the dictionaries of resultwe will be stored 
df_list = []
# Liste frame where  tweet_id of the errors  will be stored 
error_list = []

# Calculate the time of excution
start = time.time()

# Get the tweet object for all the teweets in archive dataframe 
for tweet_id in archive['tweet_id']:
    try:
        page = api.get_status(tweet_id, tweet_mode = 'extended')
        # Print one page to look at the structure of the returned file
        # and the names of attributes
        # print(json.dumps(page, indent = 4))
        #break
        
        favorites = page['favorite_count'] # How many favorites the tweet had
        retweets = page['retweet_count'] # Count of the retweet
        user_followers = page['user']['followers_count'] # How many followers the user had
        user_favourites = page['user']['favourites_count'] # How many favorites the user had
        date_time = page['created_at'] # The date and time of the creation
        
        df_list.append({'tweet_id': int(tweet_id),
                        'favorites': int(favorites),
                        'retweets': int(retweets),
                        'user_followers': int(user_followers),
                        'user_favourites': int(user_favourites),
                        'date_time': pd.to_datetime(date_time)})
    
    # Catch the exceptions of the TweepError
    except Exception as e:
        print(str(tweet_id)+ " _ " + str(e))
        error_list.append(tweet_id)

# Calculate the time of excution
end = time.time()
print(end - start)
# 888202515573088257 _ [{'code': 144, 'message': 'No status found with that ID.'}]
# 873697596434513921 _ [{'code': 144, 'message': 'No status found with that ID.'}]
# 869988702071779329 _ [{'code': 144, 'message': 'No status found with that ID.'}]
# 861769973181624320 _ [{'code': 144, 'message': 'No status found with that ID.'}]
# 842892208864923648 _ [{'code': 144, 'message': 'No status found with that ID.'}]
# 802247111496568832 _ [{'code': 144, 'message': 'No status found with that ID.'}]
# 775096608509886464 _ [{'code': 144, 'message': 'No status found with that ID.'}]
# Rate limit reached. Sleeping for: 212
# Rate limit reached. Sleeping for: 532
# 1980.119999885559


888202515573088257 _ [{'code': 144, 'message': 'No status found with that ID.'}]
873697596434513921 _ [{'code': 144, 'message': 'No status found with that ID.'}]
872668790621863937 _ [{'code': 144, 'message': 'No status found with that ID.'}]
869988702071779329 _ [{'code': 144, 'message': 'No status found with that ID.'}]
866816280283807744 _ [{'code': 144, 'message': 'No status found with that ID.'}]
861769973181624320 _ [{'code': 144, 'message': 'No status found with that ID.'}]
845459076796616705 _ [{'code': 144, 'message': 'No status found with that ID.'}]
842892208864923648 _ [{'code': 144, 'message': 'No status found with that ID.'}]
837012587749474308 _ [{'code': 144, 'message': 'No status found with that ID.'}]
827228250799742977 _ [{'code': 144, 'message': 'No status found with that ID.'}]
812747805718642688 _ [{'code': 144, 'message': 'No status found with that ID.'}]
802247111496568832 _ [{'code': 144, 'message': 'No status found with that ID.'}]
775096608509886464 _ [{'code

Rate limit reached. Sleeping for: 368


754011816964026368 _ [{'code': 144, 'message': 'No status found with that ID.'}]
680055455951884288 _ [{'code': 144, 'message': 'No status found with that ID.'}]


Rate limit reached. Sleeping for: 375


2139.599962234497


In [52]:
# lengh of the result
print("The lengh of the result", len(df_list))
# The tweet_id of the errors
print("The lengh of the errors", len(error_list))

The lengh of the result 2340
The lengh of the errors 16


According to results above:
I reached the limit of the tweepy API twice but wait_on_rate_limit automatically wait for rate limits to replenish and wait_on_rate_limit_notify print a notification when Tweepy is waiting
The total time was about 2373 seconds.
I get 2340 tweet_id correctly with 16 errors.

In [53]:
# The operation is repeated for the tweet_ids that was not get the result to df_list
ee_list = []
for e in error_list:
    try:
        favorites = page['favorite_count']
        retweets = page['retweet_count']
        user_followers = page['user']['followers_count']
        user_favourites = page['user']['favourites_count']
        date_time = page['created_at']
        
        df_list.append({'tweet_id': int(tweet_id),
                        'favorites': int(favorites),
                        'retweets': int(retweets),
                        'user_followers': int(user_followers),
                        'user_favourites': int(user_favourites),
                        'date_time': pd.to_datetime(date_time)})
        
    except Exception:
        print(str(tweet_id)+ " _ " + str(e))
        ee_list.append(e)

In [54]:
# The 16 errors saved in the list
# lengh of the result
print("The lengh of the result after Querying the errors separately", len(df_list))

The lengh of the result after Querying the errors separately 2356


In [55]:
# Create DataFrames from list of dictionaries
json_tweets = pd.DataFrame(df_list, columns = ['tweet_id', 'favorites', 'retweets',
                                              'user_followers', 'user_favourites', 'date_time'])

# Save the dataFrame in file
json_tweets.to_csv('tweet_json.txt', encoding = 'utf-8', index=False)

In [56]:


# Read the saved tweet_json.txt file into a dataframe
json_tweets = pd.read_csv('tweet_json.txt', encoding = 'utf-8')

The file tweet_json.txt successfully saved in working directory contains the result of the API Querying and stored in json_tweets data frame.

Gather step summary:
Gathering is the first step of data wrangling process. 
Obtaining data from different resources.
1- Getting data from an existing file (twitter-archive-enhanced.csv), and Reading from csv file using pandas.
2- Downloading a file from the internet (image-predictions.tsv), and Downloading file using requests. 
3- uerying twitter API (tweet_json.txt) Get JSON object of all the tweet_ids using Tweepy 
Importing that data into programming environment (Jupyter Notebook).

 # Assess

After finishing the first step which is gathering data, assess data will be the next step to asses them visually and programmatically for quality and tidiness issues. I will detect and document quality and tidiness issues.

In [57]:
# Print archive dataset to assess it visually
archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [58]:
json_tweets

tweet_id  favorites  retweets  user_followers  \
0     892420643555336193      37982      8298         7690755   
1     892177421306343426      32610      6126         7690755   
2     891815181378084864      24557      4056         7690755   
3     891689557279858688      41310      8438         7690755   
4     891327558926688256      39504      9138         7690755   
5     891087950875897856      19851      3042         7690755   
6     890971913173991426      11599      2013         7690755   
7     890729181411237888      64060     18402         7690755   
8     890609185150312448      27263      4176         7690755   
9     890240255349198849      31284      7198         7690755   
10    890006608113172480      30056      7154         7690755   
11    889880896479866881      27262      4856         7690755   
12    889665388333682689      47169      9795         7690755   
13    889638837579907072      26612      4417         7690755   
14    889531135344209921      14818      2193         7690755   
15    889278841981685760      24750      5238         7690755   
16    888917238123831296      28535      4397         7690755   
17    888804989199671297      25059      4182         7690755   
18    888554962724278272      19432      3461         7690755   
19    888078434458587136      21326      3409         7690755   
20    887705289381826560      29579      5265         7690755   
21    887517139158093824      45358     11405         7690756   
22    887473957103951883      67652     17742         7690756   
23    887343217045368832      33013     10161         7690756   
24    887101392804085760      29977      5816         7690756   
25    886983233522544640      34441      7585         7690756   
26    886736880519319552      11823      3199         7690756   
27    886680336477933568      21993      4357         7690756   
28    886366144734445568      20780      3127         7690756   
29    886267009285017600        116         4         7690756   
...                  ...        ...       ...             ...   
2326  666082916733198337        114        44         7690882   
2327  666073100786774016        315       159         7690883   
2328  666071193221509120        143        59         7690883   
2329  666063827256086533        466       213         7690883   
2330  666058600524156928        109        57         7690883   
2331  666057090499244032        290       139         7690883   
2332  666055525042405380        429       236         7690883   
2333  666051853826850816       1202       834         7690883   
2334  666050758794694657        130        59         7690883   
2335  666049248165822465        106        42         7690883   
2336  666044226329800704        293       139         7690883   
2337  666033412701032449        123        43         7690883   
2338  666029285002620928        126        47         7690883   
2339  666020888022790149       2533       500         7690883   
2340  666020888022790149       2533       500         7690883   
2341  666020888022790149       2533       500         7690883   
2342  666020888022790149       2533       500         7690883   
2343  666020888022790149       2533       500         7690883   
2344  666020888022790149       2533       500         7690883   
2345  666020888022790149       2533       500         7690883   
2346  666020888022790149       2533       500         7690883   
2347  666020888022790149       2533       500         7690883   
2348  666020888022790149       2533       500         7690883   
2349  666020888022790149       2533       500         7690883   
2350  666020888022790149       2533       500         7690883   
2351  666020888022790149       2533       500         7690883   
2352  666020888022790149       2533       500         7690883   
2353  666020888022790149       2533       500         7690883   
2354  666020888022790149       2533       500         7690883   
2355  666020888022790149       2533       500      

In [59]:
images

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [60]:
# Assessing the data programmaticaly
archive.info()
archive.describe()
archive['rating_numerator'].value_counts()
archive['rating_denominator'].value_counts()
archive['name'].value_counts()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

None         745
a             55
Charlie       12
Cooper        11
Oliver        11
Lucy          11
Penny         10
Lola          10
Tucker        10
Bo             9
Winston        9
the            8
Sadie          8
Buddy          7
an             7
Daisy          7
Bailey         7
Toby           7
Jax            6
Dave           6
Leo            6
Milo           6
Jack           6
Koda           6
Oscar          6
Rusty          6
Bella          6
Scout          6
Stanley        6
Finn           5
            ... 
Cedrick        1
Wafer          1
Keet           1
Bowie          1
Chef           1
Julio          1
Mosby          1
Covach         1
Skittle        1
Bodie          1
Strudel        1
Grizzwald      1
Remus          1
Carbon         1
Rumble         1
Wesley         1
Lambeau        1
BeBe           1
Rinna          1
DonDon         1
Jareld         1
Jockson        1
Heinrich       1
Ralphé         1
Ralph          1
Alejandro      1
Dido           1
Clarkus       

In [61]:
images.info()
images['jpg_url'].value_counts()
images[images['jpg_url'] == 'https://pbs.twimg.com/media/CZhn-QAWwAASQan.jpg']

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


tweet_id                                          jpg_url  \
800   691416866452082688  https://pbs.twimg.com/media/CZhn-QAWwAASQan.jpg   
1624  803692223237865472  https://pbs.twimg.com/media/CZhn-QAWwAASQan.jpg   

      img_num                p1   p1_conf  p1_dog             p2   p2_conf  \
800         1  Lakeland_terrier  0.530104    True  Irish_terrier  0.197314   
1624        1  Lakeland_terrier  0.530104    True  Irish_terrier  0.197314   

      p2_dog        p3   p3_conf  p3_dog  
800     True  Airedale  0.082515    True  
1624    True  Airedale  0.082515    True

In [62]:
json_tweets
json_tweets.info()
# json_tweets['tweet_id'].value_counts() count tweet_ids
# json_tweets['user_followers'].value_counts() check if querying the use_followers had a meaning

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 6 columns):
tweet_id           2356 non-null int64
favorites          2356 non-null int64
retweets           2356 non-null int64
user_followers     2356 non-null int64
user_favourites    2356 non-null int64
date_time          2356 non-null object
dtypes: int64(5), object(1)
memory usage: 110.5+ KB


In [63]:
sum(archive['tweet_id'].duplicated())

0

In [64]:
sum(images.jpg_url.duplicated())

66

 # Quality

Completeness, validity, accuracy, consistency (content issues).

### Archive Dataset

1- Convert timestamp column to datetime with day, month and year

2- Columns that will not be used for analysis must be deleted

3- Names of dogs are miss labelled, mispelled or missing.

4- Combine the 'doggo', 'floofer', 'pupper' and 'puppo' columns into one column named 'dog_stage'. 

5- tweet_id is int not string.

### Images Dataset

1- There are 66 images jpg_url duplicated they must be dropped.

2- tweet_id is int not string

3- Columns that will not be used for analysis must be deleted.

4- There are missing values from images dataset.

5- Some tweets are have 2 different tweet_id one refer to the other.

### Json_Tweets Dataset

1- Rename the id column to "tweet_id" to match the other 2 datasets.

2- 176 records have a retweeted_status, will need to be deleted.


# Tidiness

### Untidy data structural issues

1- All tables should be part of one dataset.

2- Four types are in one column(doggo, floofer, pupper and puppo).


# Clean

Cleaning data is the third step of data wrangling steps. It is to fix quality and tidiness issues that were identified in the assess step.

In [94]:
# make a copy of dataFrames before cleaning
archive = archive.copy()
images = images.copy()
json_tweets = json_tweets.copy()

In [95]:
# merging the tables into on dataframe and save the result in file as backup
twitter_dog = pd.merge(archive, images, how = 'left', on = ['tweet_id'] )
twitter_dog = pd.merge(twitter_dog, json_tweets, how = 'left', on = ['tweet_id'])
twitter_dog.to_csv('twitter_dog.csv', encoding = 'utf-8')
twitter_dog.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2372 entries, 0 to 2371
Data columns (total 33 columns):
tweet_id                      2372 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2372 non-null object
source                        2372 non-null object
text                          2372 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2313 non-null object
rating_numerator              2372 non-null int64
rating_denominator            2372 non-null int64
name                          2372 non-null object
doggo                         2372 non-null object
floofer                       2372 non-null object
pupper                        2372 non-null object
puppo                         2372 non-null object
jpg_url                       20

In [96]:
twitter_dog

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN                  NaN   
2350  666050758794694657                    NaN                  NaN   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   
2356  666020888022790149                    NaN                  NaN   
2357  666020888022790149                    NaN                  NaN   
2358  666020888022790149                    NaN                  NaN   
2359  666020888022790149                    NaN                  NaN   
2360  666020888022790149                    NaN                  NaN   
2361  666020888022790149                    NaN                  NaN   
2362  666020888022790149                    NaN                  NaN   
2363  666020888022790149                    NaN                  NaN   
2364  666020888022790149                    NaN                  NaN   
2365  666020888022790149                    NaN         

###  Archive 1
#####  Define
Convert timestamp column to datetime with day, month and year.

#####  Code

In [80]:
archive['timestamp'] = pd.to_datetime(archive['timestamp'])

#extract year, month and day to new columns
archive['year'] = archive['timestamp'].dt.year
archive['month'] = archive['timestamp'].dt.month
archive['day'] = archive['timestamp'].dt.day

#drop timestamp column
archive = archive.drop('timestamp', 1)

##### Test

In [81]:
archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

 ### Archive 2
  ##### Define 
 Delete columns that will not be used for analysis
 ##### Code

In [69]:
# Delete columns no needed
archive = archive.drop(['source',
                                                    'in_reply_to_status_id',
                                                    'in_reply_to_user_id',
                                                    'retweeted_status_id',
                                                    'retweeted_status_user_id', 
                                                    'retweeted_status_timestamp', 
                                                    'expanded_urls'], 1)

##### Test

In [84]:
list(archive)

['tweet_id',
 'in_reply_to_status_id',
 'in_reply_to_user_id',
 'source',
 'text',
 'retweeted_status_id',
 'retweeted_status_user_id',
 'retweeted_status_timestamp',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'doggo',
 'floofer',
 'pupper',
 'puppo',
 'year',
 'month',
 'day']

### Archive 3
##### Define
Replace names of dogs are miss labelled, mispelled or missing with the correct name or None.
##### Code

In [83]:


# Missing Data
# replacing faulty names with None value or corrected Name
archive['name'].replace('the', 'None', inplace=True)
archive['name'].replace("light",'None', inplace=True)
archive['name'].replace("life",'None', inplace=True)
archive['name'].replace("an",'None', inplace=True)
archive['name'].replace("a",'None', inplace=True)
archive['name'].replace("by",'None', inplace=True)
archive['name'].replace("actually",'None', inplace=True)
archive['name'].replace("just",'None', inplace=True)
archive['name'].replace("getting",'None', inplace=True) 
archive['name'].replace("infuriating",'None', inplace=True) 
archive['name'].replace("old",'None', inplace=True) 
archive['name'].replace("all",'None', inplace=True) 
archive['name'].replace("this",'None', inplace=True) 
archive['name'].replace("very",'None', inplace=True) 
archive['name'].replace("mad",'None', inplace=True) 
archive['name'].replace("not",'None', inplace=True)
archive['name'].replace("one",'None', inplace=True)
archive['name'].replace("my",'None', inplace=True)
archive['name'].replace("O","O'Malley", inplace=True)
archive['name'].replace("quite","None", inplace=True)
archive['name'].replace("such","None", inplace=True)

##### Test

In [85]:
archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

### Archive 4
##### Define 
Combine the 'doggo', 'floofer', 'pupper' and 'puppo' columns into one column named 'dog_stage'. 
##### Code

In [97]:
# Select the columns to melt and to remain
columns_to_melt = ['doggo', 'floofer', 'pupper', 'puppo']
columns_to_stay = [x for x in twitter_dog.columns.tolist() if x not in columns_to_melt]

# Mlet the the columns into values
twitter_dog = pd.melt(twitter_dog, id_vars = columns_to_stay, value_vars = columns_to_melt, 
                         var_name = 'text', value_name = 'dog_stage')

# Delete column 'stages'
twitter_dog = twitter_dog.drop('text', 1)

# Filter for unique values then remove duplicate values based on 'dog_stage' values

# This part for test *
print(twitter_dog.dog_stage.value_counts())

twitter_dog = twitter_dog.sort_values('dog_stage').drop_duplicates('tweet_id', keep = 'last')

# This part for test
print(twitter_dog.dog_stage.value_counts())
print(len(twitter_dog))

None       9094
pupper      257
doggo        97
puppo        30
floofer      10
Name: dog_stage, dtype: int64
None       1976
pupper      257
doggo        83
puppo        30
floofer      10
Name: dog_stage, dtype: int64
2356


##### Test 

In [98]:
twitter_dog

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2274  667455448082227200                    NaN                  NaN   
2275  667453023279554560                    NaN                  NaN   
2276  667443425659232256                    NaN                  NaN   
2277  667437278097252352                    NaN                  NaN   
2278  667435689202614272                    NaN                  NaN   
2279  667405339315146752                    NaN                  NaN   
2273  667470559035432960                    NaN                  NaN   
2280  667393430834667520                    NaN                  NaN   
2282  667211855547486208                    NaN                  NaN   
2283  667200525029539841                    NaN                  NaN   
2284  667192066997374976                    NaN                  NaN   
2285  667188689915760640                    NaN                  NaN   
2286  667182792070062081                    NaN                  NaN   
2287  667177989038297088                    NaN                  NaN   
2281  667369227918143488                    NaN                  NaN   
2272  667491009379606528                    NaN                  NaN   
2271  667495797102141441                    NaN                  NaN   
2270  667502640335572993                    NaN                  NaN   
2255  667773195014021121                    NaN                  NaN   
2256  667766675769573376                    NaN                  NaN   
2257  667728196545200128                    NaN                  NaN   
2258  667724302356258817                    NaN                  NaN   
2259  667550904950915073                    NaN                  NaN   
2260  667550882905632768                    NaN                  NaN   
2261  667549055577362432                    NaN                  NaN   
2262  667546741521195010                    NaN                  NaN   
2263  667544320556335104                    NaN                  NaN   
2264  667538891197542400                    NaN                  NaN   
2265  667534815156183040                    NaN                  NaN   
2266  667530908589760512                    NaN                  NaN   
...                  ...                    ...                  ...   
7511  825535076884762624                    NaN                  NaN   
7307  855851453814013952                    NaN                  NaN   
7662  805823200554876929                    NaN                  NaN   
7130  889531135344209921                    NaN                  NaN   
7759  793195938047070209                    NaN                  NaN   
7128  889665388333682689                    NaN                  NaN   
8151  744995568523612160                    NaN                  NaN   
7513  825120256414846976                    NaN                  NaN   
8063  752519690950500352                    NaN                  NaN   
7284  859607811541651456                    NaN                  NaN   
7829  784183165795655680                    NaN                  NaN   
7585  816829038950027264                    NaN                  NaN   
8077  751132876104687617                    NaN                  NaN   
7210  874012996292530176                    NaN                  NaN   
7817  786051337297522688           7.727430e+17         7.305050e+17   
7245  867421006826221569                    NaN                  NaN   
8199  738537504001953792                    NaN                  NaN   
7683  802239329049477120                    NaN                  NaN   
7514  825026590719483904                    NaN                  NaN   
7555  819952236453363712                    NaN                  NaN   
7805  787717603741622272                    NaN                  NaN   
8038  756275833623502848                    NaN                  NaN   
7187  878776093423087618                    NaN                  NaN   
7852  780931614150983680                    NaN         

### Archive 5
##### Define 
Convert tweet_id column to String.
##### Code

In [99]:
archive.tweet_id = archive.tweet_id.astype(str)

##### Test

In [100]:
archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

### Images 1
##### Define
Delete the 66 duplicated jpg_url
##### Code

In [102]:
images = images.drop_duplicates(subset=['jpg_url'], keep='last')

##### Test

In [103]:
sum(images['jpg_url'].duplicated())

0

### Images 2
##### Define
Convert tweet_id column to String.
##### Code

In [105]:
images.tweet_id = images.tweet_id.astype(str)

##### Test

In [106]:
images

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2044  886258384151887873    https://pbs.twimg.com/media/DEyfTG4UMAE4aE9.jpg   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

### Json_tweets 1
##### Define
Rename the id column to "tweet_id" to match the other 2 datasets
##### Code

In [109]:
json_tweets.rename(columns={'id': 'tweet_id'}, inplace=True)
json_tweets = json_tweets.tweet_id.astype(str)

##### Test

In [110]:
json_tweets

0       892420643555336193
1       892177421306343426
2       891815181378084864
3       891689557279858688
4       891327558926688256
5       891087950875897856
6       890971913173991426
7       890729181411237888
8       890609185150312448
9       890240255349198849
10      890006608113172480
11      889880896479866881
12      889665388333682689
13      889638837579907072
14      889531135344209921
15      889278841981685760
16      888917238123831296
17      888804989199671297
18      888554962724278272
19      888078434458587136
20      887705289381826560
21      887517139158093824
22      887473957103951883
23      887343217045368832
24      887101392804085760
25      886983233522544640
26      886736880519319552
27      886680336477933568
28      886366144734445568
29      886267009285017600
               ...        
2326    666082916733198337
2327    666073100786774016
2328    666071193221509120
2329    666063827256086533
2330    666058600524156928
2331    666057090499244032
2

Melt the 'doggo', 'floofer', 'pupper' and 'puppo' columns into one column named 'dog_stage

In [73]:
# Select the columns to melt and to remain
columns_to_melt = ['doggo', 'floofer', 'pupper', 'puppo']
columns_to_stay = [x for x in twitter_dog.columns.tolist() if x not in columns_to_melt]

# Mlet the the columns into values
twitter_dog = pd.melt(twitter_dog, id_vars = columns_to_stay, value_vars = columns_to_melt, 
                         var_name = 'text', value_name = 'dog_stage')

# Delete column 'stages'
twitter_dog = twitter_dog.drop('text', 1)

# Filter for unique values then remove duplicate values based on 'dog_stage' values

# This part for test *
print(twitter_dog.dog_stage.value_counts())

twitter_dog = twitter_dog.sort_values('dog_stage').drop_duplicates('tweet_id', keep = 'last')

# This part for test
print(twitter_dog.dog_stage.value_counts())
print(len(twitter_dog))

None       9094
pupper      257
doggo        97
puppo        30
floofer      10
Name: dog_stage, dtype: int64
None       1976
pupper      257
doggo        83
puppo        30
floofer      10
Name: dog_stage, dtype: int64
2356


In [74]:
twitter_dog

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2274  667455448082227200                    NaN                  NaN   
2275  667453023279554560                    NaN                  NaN   
2276  667443425659232256                    NaN                  NaN   
2277  667437278097252352                    NaN                  NaN   
2278  667435689202614272                    NaN                  NaN   
2279  667405339315146752                    NaN                  NaN   
2273  667470559035432960                    NaN                  NaN   
2280  667393430834667520                    NaN                  NaN   
2282  667211855547486208                    NaN                  NaN   
2283  667200525029539841                    NaN                  NaN   
2284  667192066997374976                    NaN                  NaN   
2285  667188689915760640                    NaN                  NaN   
2286  667182792070062081                    NaN                  NaN   
2287  667177989038297088                    NaN                  NaN   
2281  667369227918143488                    NaN                  NaN   
2272  667491009379606528                    NaN                  NaN   
2271  667495797102141441                    NaN                  NaN   
2270  667502640335572993                    NaN                  NaN   
2255  667773195014021121                    NaN                  NaN   
2256  667766675769573376                    NaN                  NaN   
2257  667728196545200128                    NaN                  NaN   
2258  667724302356258817                    NaN                  NaN   
2259  667550904950915073                    NaN                  NaN   
2260  667550882905632768                    NaN                  NaN   
2261  667549055577362432                    NaN                  NaN   
2262  667546741521195010                    NaN                  NaN   
2263  667544320556335104                    NaN                  NaN   
2264  667538891197542400                    NaN                  NaN   
2265  667534815156183040                    NaN                  NaN   
2266  667530908589760512                    NaN                  NaN   
...                  ...                    ...                  ...   
7511  825535076884762624                    NaN                  NaN   
7307  855851453814013952                    NaN                  NaN   
7662  805823200554876929                    NaN                  NaN   
7130  889531135344209921                    NaN                  NaN   
7759  793195938047070209                    NaN                  NaN   
7128  889665388333682689                    NaN                  NaN   
8151  744995568523612160                    NaN                  NaN   
7513  825120256414846976                    NaN                  NaN   
8063  752519690950500352                    NaN                  NaN   
7284  859607811541651456                    NaN                  NaN   
7829  784183165795655680                    NaN                  NaN   
7585  816829038950027264                    NaN                  NaN   
8077  751132876104687617                    NaN                  NaN   
7210  874012996292530176                    NaN                  NaN   
7817  786051337297522688           7.727430e+17         7.305050e+17   
7245  867421006826221569                    NaN                  NaN   
8199  738537504001953792                    NaN                  NaN   
7683  802239329049477120                    NaN                  NaN   
7514  825026590719483904                    NaN                  NaN   
7555  819952236453363712                    NaN                  NaN   
7805  787717603741622272                    NaN                  NaN   
8038  756275833623502848                    NaN                  NaN   
7187  878776093423087618                    NaN                  NaN   
7852  780931614150983680                    NaN         

In [75]:
archive

tweet_id                                               text  \
0     892420643555336193  This is Phineas. He's a mystical boy. Only eve...   
1     892177421306343426  This is Tilly. She's just checking pup on you....   
2     891815181378084864  This is Archie. He is a rare Norwegian Pouncin...   
3     891689557279858688  This is Darla. She commenced a snooze mid meal...   
4     891327558926688256  This is Franklin. He would like you to stop ca...   
5     891087950875897856  Here we have a majestic great white breaching ...   
6     890971913173991426  Meet Jax. He enjoys ice cream so much he gets ...   
7     890729181411237888  When you watch your owner call another dog a g...   
8     890609185150312448  This is Zoey. She doesn't want to be one of th...   
9     890240255349198849  This is Cassie. She is a college pup. Studying...   
10    890006608113172480  This is Koda. He is a South Australian decksha...   
11    889880896479866881  This is Bruno. He is a service shark. Only get...   
12    889665388333682689  Here's a puppo that seems to be on the fence a...   
13    889638837579907072  This is Ted. He does his best. Sometimes that'...   
14    889531135344209921  This is Stuart. He's sporting his favorite fan...   
15    889278841981685760  This is Oliver. You're witnessing one of his m...   
16    888917238123831296  This is Jim. He found a fren. Taught him how t...   
17    888804989199671297  This is Zeke. He has a new stick. Very proud o...   
18    888554962724278272  This is Ralphus. He's powering up. Attempting ...   
19    888202515573088257  RT @dog_rates: This is Canela. She attempted s...   
20    888078434458587136  This is Gerald. He was just told he didn't get...   
21    887705289381826560  This is Jeffrey. He has a monopoly on the pool...   
22    887517139158093824  I've yet to rate a Venezuelan Hover Wiener. Th...   
23    887473957103951883  This is Canela. She attempted some fancy porch...   
24    887343217045368832  You may not have known you needed to see this ...   
25    887101392804085760  This... is a Jubilant Antarctic House Bear. We...   
26    886983233522544640  This is Maya. She's very shy. Rarely leaves he...   
27    886736880519319552  This is Mingus. He's a wonderful father to his...   
28    886680336477933568  This is Derek. He's late for a dog meeting. 13...   
29    886366144734445568  This is Roscoe. Another pupper fallen victim t...   
...                  ...                                                ...   
2326  666411507551481857  This is quite the dog. Gets really excited whe...   
2327  666407126856765440  This is a southern Vesuvius bumblegruff. Can d...   
2328  666396247373291520  Oh goodness. A super rare northeast Qdoba kang...   
2329  666373753744588802  Those are sunglasses and a jean jacket. 11/10 ...   
2330  666362758909284353  Unique dog here. Very small. Lives in containe...   
2331  666353288456101888  Here we have a mixed Asiago from the Galápagos...   
2332  666345417576210432  Look at this jokester thinking seat belt laws ...   
2333  666337882303524864  This is an extremely rare horned Parthenon. No...   
2334  666293911632134144  This is a funny dog. Weird toes. Won't come do...   
2335  666287406224695296  This is an Albanian 3 1/2 legged  Episcopalian...   
2336  666273097616637952     Can take selfies 11/10 https://t.co/ws2AMaNwPW   
2337  666268910803644416  Very concerned about fellow dog trapped in com...   
2338  666104133288665088  Not familiar with this breed. No tail (weird)....   
2339  666102155909144576  Oh my. Here you are seeing an Adobe Setter giv...   
2340  666099513787052032  Can stand on stump for what seems like a while...   
2341  666094000022159362  This appears to be a Mongolian Presbyterian mi...   
2342  666082916733198337  Here we have a well-established sunblockerspan...   
2343  666073100786774016  Let's hope this flight isn't Malaysian (lol). ...   
2344  666071193221509120  Here we have a northern speckled Rhodode

In [111]:
#Store the clean dataset into dataframe
twitter_dog.to_csv('twitter_archive_master.csv', encoding='utf-8',index=False)